OpenRA.Mods.D2k/UtilityCommands/D2kMapImporter.cs

```diff
@@ -0,0 +1,419 @@
+
+				terrainTiles.Add(GetTile(data));
+				if (specialData != 0)
+					specialValueForTileIndex.Add(index / 4 - 1, specialData);
+			}
+		}
+
+		void Initialize(string mapFile)
+		{
+			mapSize = new Size(stream.ReadUInt16(), stream.ReadUInt16());
+
+			tileSet = rules.TileSets["ARRAKIS"];
+			map = Map.FromTileset(tileSet);
+			map.Title = Path.GetFileNameWithoutExtension(mapFile);
+			map.Author = "Westwood Studios";
+			map.MapSize.X = mapSize.Width + 2 * MapCordonWidth;
+			map.MapSize.Y = mapSize.Height + 2 * MapCordonWidth;
+			map.Bounds = new Rectangle(MapCordonWidth, MapCordonWidth, mapSize.Width, mapSize.Height);
+
+			map.Smudges = Exts.Lazy(() => new List<SmudgeReference>());
+			map.Actors = Exts.Lazy(() => new Dictionary<string, ActorReference>());
+			map.MapResources = Exts.Lazy(() => new CellLayer<ResourceTile>(TileShape.Rectangle, new Size(map.MapSize.X, map.MapSize.Y)));
+			map.MapTiles = Exts.Lazy(() => new CellLayer<TerrainTile>(TileShape.Rectangle, new Size(map.MapSize.X, map.MapSize.Y)));
+
+			map.Options = new MapOptions();
+
+			terrainTiles = new List<TerrainTile>();
+			specialValueForTileIndex = new Dictionary<int, int>();
+
+			// Get all templates from the tileset YAML file that have at least one frame and an Image property corresponding to the requested tileset
+			// Each frame is a tile from the Dune 2000 tileset files, with the Frame ID being the index of the tile in the original file
+			tileSetsFromYaml = tileSet.Templates.Where(t => t.Value.Frames != null
+				&& t.Value.Image.ToLower() == tilesetName.ToLower()).Select(ts => ts.Value).ToList();
+		}
+
+		void FillMap()
+		{
+			for (var y = 0; y < map.MapSize.Y; y++)
+				for (var x = 0; x < map.MapSize.X; x++)
+				{
+					var locationOnMap = new CPos(x, y);
+
+					// Hack for a "clear" tile
+					var indexOnMap = 0;
+
+					if (map.Contains(locationOnMap))
+						indexOnMap = (y - MapCordonWidth) * mapSize.Width + (x - MapCordonWidth); // +1;
+
+					if (indexOnMap < 0 || indexOnMap >= terrainTiles.Count)
+						indexOnMap = 0;
+
+					map.MapTiles.Value[locationOnMap] = terrainTiles[indexOnMap];
+
+					// Tile special values
+					if (!specialValueForTileIndex.ContainsKey(indexOnMap))
+						continue;
+
+					var tileSpecialInfo = specialValueForTileIndex[indexOnMap];
+
+					// Spice
+					if (tileSpecialInfo == 1)
+						map.MapResources.Value[locationOnMap] = new ResourceTile(1, 1);
+					if (tileSpecialInfo == 2)
+						map.MapResources.Value[locationOnMap] = new ResourceTile(1, 2);
+
+					// Actors
+					if (actorDataByActorCode.ContainsKey(tileSpecialInfo))
+					{
+						var kvp = actorDataByActorCode[tileSpecialInfo];
+						if (!rules.Actors.ContainsKey(kvp.First.ToLower()))
+							throw new InvalidOperationException("Actor with name {0} could not be found in the rules YAML file!".F(kvp.First));
+
+						var a = new ActorReference(kvp.First)
+						{
+							new LocationInit(locationOnMap),
+							new OwnerInit(kvp.Second)
```

Comment:
Ah, my bad with the above comment, didn't look hard enough at what you were commenting on.
The `Actors` part will only be viable for a MAP file that comes together with a MIS file anyway (so only for missions - official or custom). And my plan was to make the utility read both of those simultaneously if they are available in order to populate the map.yaml with the appropriate player definitions.

Like I said earlier, right now it works for skirmish/multiplayer maps, but not for mission maps (in the Dune 2000 sense), where you have to open the map.yaml and fix it yourself.
This is just one of the reasons the MIS file reader is important to implement.